### PR TITLE
lightning: support importing timestamp from Hive parquet

### LINKED ISSUE #37536: lightning: ts discrepancy between parquet file from Hive and imported data
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. export parquet file by:
```
create table (...) stored as parquet location '....'
```
2. import parquet file with lightning using local-backend
3. read parquet file by python and compare the result with data in TiDB

### 2. What did you expect to see? (Required)
No diff
### 3. What did you see instead (Required)
result in parquet:
image

result in TiDB:
image

### 4. What is your TiDB version? (Required)

TiDB: 36a9810441ca0e496cbd22064af274b3be771081
Lightning: e8daf1ad7d0a4dabf2bcb53f4b0850412a02944d

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

Issue Number: close #37536 

Problem Summary:

### What is changed and how it works?
The timestamp exported from Hive will be encoded as INT96 which is natively supported by neither parquet-go nor Golang itself. Therefore, this type of timestamp will be stored as strings in lightning. Because we cannot extract timestamps from this byte string, the import result is kind of fallacious.

This PR is about to identify this type of timestamp and decode true time from it.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 1]
```diff
@@ -446,6 +449,13 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
 	if meta.LogicalType != nil && meta.LogicalType.DECIMAL != nil {
 		v = binaryToDecimalStr([]byte(v), int(meta.LogicalType.DECIMAL.Scale))
 	}
+	if meta.Type != nil && *meta.Type == parquet.Type_INT96 && len([]byte(v)) == 12 {
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:452]:
```suggestion
	if meta.Type != nil && *meta.Type == parquet.Type_INT96 && len(v) == 12 {
```

Or maybe use 96/8 to replace 12 😂

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 2]
```diff
@@ -446,6 +449,13 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
...
+		ts := int96ToTime([]byte(v))
+		ts = ts.UTC()
+		tsStr := ts.Format("2006-01-02 15:04:05.999999Z")
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:455]: why do we choose this layout?

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 3]
```diff
@@ -566,3 +576,14 @@ func (*ParquetParser) SetColumns(_ []string) {
 func (pp *ParquetParser) SetLogger(l log.Logger) {
 	pp.logger = l
 }
+
+func jdToTime(jd int32, nsec int64) time.Time {
+	sec := int64(jd-jan011970) * secPerDay
+	return time.Unix(sec, int64(nsec))
+}
+
+func int96ToTime(parquetDate []byte) time.Time {
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:585]: can you refer to a parquet encoding for timestamp as comment?

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 4]
```diff
@@ -26,6 +27,8 @@ const (
 	// if a parquet if small than this threshold, parquet will load the whole file in a byte slice to
 	// optimize the read performance
 	smallParquetFileThreshold = 256 * 1024 * 1024
+	jan011970                 = 2440588
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:30]: what's the meaning of 2440588

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 5]
```diff
@@ -566,3 +576,14 @@ func (*ParquetParser) SetColumns(_ []string) {
 func (pp *ParquetParser) SetLogger(l log.Logger) {
 	pp.logger = l
 }
+
+func jdToTime(jd int32, nsec int64) time.Time {
+	sec := int64(jd-jan011970) * secPerDay
+	return time.Unix(sec, int64(nsec))
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:582]:
```suggestion
	return time.Unix(sec, nsec)
```

And can you add unit tests for nsec outside [0, 999,999,999]? (see the comment of `time.Unix`)

Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:582]: Will take a look that this later.

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 4 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:30]: The decoding of the INT96 timestamp refers to a parquet library called [parquet-go](https://github.com/fraugster/parquet-go). Will paste some comments here tomorrow.

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 3 above)
Comment (lance6716) [reply br/pkg/lightning/mydump/parquet_parser.go:585]: https://github.com/apache/spark/blob/d66a4e82eceb89a274edeb22c2fb4384bed5078b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala#L171-L178

maybe this is enough

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/lightning/mydump/parquet_parser.go:452]: what's the value of LogicalType if enter this branch

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 6]
```diff
@@ -446,6 +449,13 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
 	if meta.LogicalType != nil && meta.LogicalType.DECIMAL != nil {
 		v = binaryToDecimalStr([]byte(v), int(meta.LogicalType.DECIMAL.Scale))
 	}
+	if meta.Type != nil && *meta.Type == parquet.Type_INT96 && len([]byte(v)) == 12 {
+		ts := int96ToTime([]byte(v))
+		ts = ts.UTC()
+		tsStr := ts.Format("2006-01-02 15:04:05.999999Z")
+		d.SetString(tsStr, "")
+		return
```
Comment (D3Hunter) [inline br/pkg/lightning/mydump/parquet_parser.go:457]:
```suggestion
		v = ts.Format("2006-01-02 15:04:05.999999Z")
```

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 1 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:452]: No logical type is set (nil).

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 2 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:455]: copy from https://github.com/pingcap/tidb/blob/ec9e43ed7e5a0b8e840da6e4dfe7695e1aac7355/br/pkg/lightning/mydump/parquet_parser.go#L512-L513

This layout is used when setting a UTC timestamp.

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 6 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 2536ce5

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 5 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 2536ce5

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 4 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 2536ce5

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 3 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:598]: 2536ce5

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 1 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 2536ce5

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 7]
```diff
@@ -26,6 +27,10 @@ const (
 	// if a parquet if small than this threshold, parquet will load the whole file in a byte slice to
 	// optimize the read performance
 	smallParquetFileThreshold = 256 * 1024 * 1024
+	// jan011970 is the start date of any julian day,
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:30]:
```suggestion
	// jan011970 is the date of unix epoch in julian day,
```

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 8]
```diff
@@ -26,6 +27,10 @@ const (
...
+	// we calcuate the offset to it to determine the date of a timestamp.
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:31]:
```suggestion
```

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 9]
```diff
@@ -446,6 +451,11 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
 	if meta.LogicalType != nil && meta.LogicalType.DECIMAL != nil {
 		v = binaryToDecimalStr([]byte(v), int(meta.LogicalType.DECIMAL.Scale))
 	}
+	if meta.Type != nil && *meta.Type == parquet.Type_INT96 && len([]byte(v)) == 96/8 {
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:454]:
```suggestion
	if meta.Type != nil && *meta.Type == parquet.Type_INT96 && len(v) == 96/8 {
```

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 10]
```diff
@@ -446,6 +451,11 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
...
+		ts := int96ToTime([]byte(v))
+		ts = ts.UTC()
+		v = ts.Format("2006-01-02 15:04:05.999999Z")
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:457]: this variable can be extracted as a constant

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 11]
```diff
@@ -566,3 +576,27 @@ func (*ParquetParser) SetColumns(_ []string) {
...
+	// See TestNsecOutSideRange.
+	return time.Unix(sec, nsec)
+}
+
+// FYI: https://github.com/apache/spark/blob/d66a4e82eceb89a274edeb22c2fb4384bed5078b/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetWriteSupport.scala#L171-L178
+// INT96 timestamp layout
+// --------------------------
+// |   64 bit   |   32 bit   |
+// ---------------------------
+// |  nano sec  |  timestamp |
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:593]:
```suggestion
// |  nano sec  |  julian day  |
```

**Review by lance6716 — APPROVED** [review]
rest lgtm

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 10 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 1f379a1

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 9 above)
Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go]: 1f379a1

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**